multiprocess: Add IPC wrapper for Mining interface

ryanofsky opened the pull request:
Add Cap'n Proto wrapper for the Mining interface introduced in #30200, and its associated types.

This PR combined with #30509 will allow a separate mining process, like the one being implemented in https://github.com/Sjors/bitcoin/pull/48, to connect to the node over IPC, and create, manage, and submit block templates. (#30437 shows another simpler demo of a process using the Mining interface.)

---

This PR is part of the [process separation project](https://github.com/bitcoin/bitcoin/issues/28722).

ryanofsky committed:
multiprocess: Add serialization code for vector<char>

ryanofsky committed:
multiprocess: Add serialization code for CTransaction

Add support for passing CTransaction and CTransactionRef types to IPC
functions.

These types can't be passed currently because IPC serialization code currently
only supports deserializing types that have an Unserialize() method, which
CTransaction does not, because it is supposed to represent immutable
transactions. Work around this by adding a CustomReadField overload that will
call CTransaction's deserialize_type constructor.

These types also can't be passed currently because serializing transactions
requires TransactionSerParams to be set. Fix this by setting TX_WITH_WITNESS as
default serialization parameters for IPC code.

ryanofsky committed:
doc: multiprocess documentation improvements

Most improvements suggested by stickies-v <stickies-v@protonmail.com>
https://github.com/bitcoin/bitcoin/pull/28978#pullrequestreview-1800375604

Omission of CustomReadMessage and CustomBuildMessage was noticed by
TheCharlatan in
https://github.com/bitcoin/bitcoin/pull/30510#discussion_r1708996040 and is
fixed here as well.

Co-authored-by: stickies-v <stickies-v@protonmail.com>

ryanofsky committed:
multiprocess: Add serialization code for BlockValidationState

Co-authored-by: TheCharlatan <seb.kung@gmail.com>

ryanofsky committed:
multiprocess: Add IPC wrapper for Mining interface

Sjors commented:
Is the idea to incorporate the interface changes proposed in #30440 and #30409 after they're merged? And in the mean time keep them in #30437?

ryanofsky commented:
> Is the idea to incorporate the interface changes proposed in #30440 and #30409 after they're merged? And in the mean time keep them in #30437?

I could add them here. Just let me know what you prefer. I am planning to rebase #30437 on top of this, too.

> Maybe also link to #30437 in the description as it's a more simple demo than [Sjors#48](https://github.com/Sjors/bitcoin/pull/48).

Thanks, added

ryanofsky commented:
Updated 4e67086e4c0babb97be30bd9fe3c3b96e258a502 -> 79cfc0d198faecec598f923e6bf83ccf8dc09bbb ([`pr/mine-types.1`](https://github.com/ryanofsky/bitcoin/commits/pr/mine-types.1) -> [`pr/mine-types.2`](https://github.com/ryanofsky/bitcoin/commits/pr/mine-types.2), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/mine-types.1..pr/mine-types.2)) to fix link error in   https://github.com/bitcoin/bitcoin/pull/30510/checks?check_run_id=27810447228

ryanofsky committed:
depends: Update libmultiprocess library for cmake headers target

This update brings in the following changes:

https://github.com/chaincodelabs/libmultiprocess/pull/107 example: Remove manual client adding
https://github.com/chaincodelabs/libmultiprocess/pull/108 doc: Add comments for socket descriptor handling when forking
https://github.com/chaincodelabs/libmultiprocess/pull/109 example: Add missing thread.join() call so example can exit cleanly
https://github.com/chaincodelabs/libmultiprocess/pull/110 cmake: add target_capnp_sources headers target

ryanofsky commented:
Was able to simplify this by dropping CustomBuildMessage/CustomReadMessage functions and tests. It now requires the latest version of the libmultiprocess library though.

Rebased 79cfc0d198faecec598f923e6bf83ccf8dc09bbb -> 6cccb436557c58e0ccd21ffe0eaf31058f5cb799 ([`pr/mine-types.2`](https://github.com/ryanofsky/bitcoin/commits/pr/mine-types.2) -> [`pr/mine-types.3`](https://github.com/ryanofsky/bitcoin/commits/pr/mine-types.3), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/mine-types.2-rebase..pr/mine-types.3)) dropping the first commit

ryanofsky reviewed (commented):
Updated 6cccb436557c58e0ccd21ffe0eaf31058f5cb799 -> c538ec69f266b51c893a374a4bb82796ede3d7cb ([`pr/mine-types.3`](https://github.com/ryanofsky/bitcoin/commits/pr/mine-types.3) -> [`pr/mine-types.4`](https://github.com/ryanofsky/bitcoin/commits/pr/mine-types.4), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/mine-types.3..pr/mine-types.4)) with review suggestions.

Thanks for the review and thoughtful questions. I wrote detailed answers in line but I will reuse the text & diagram in documentation for libmultiprocess so there's a place to point to if similar questions come up again.

TheCharlatan commented:
Thanks for all the explanations @ryanofsky , I'll take some time to digest all of this, and read some more libmultiprocess code.

ryanofsky commented:
> Thanks for all the explanations @ryanofsky , I'll take some time to digest all of this, and read some more libmultiprocess code.

Thanks! Feel free to ping me if you have any questions. it will help me learn what things to work on documentation for.

Sjors commented on src/ipc/capnp/common-types.h:
```diff
@@ -16,6 +17,14 @@
+template<typename S>
+auto Wrap(S& s)
+{
+    return ParamsStream{s, TX_WITH_WITNESS};
```
8523d19bb3f8c982c6e01a4635c98cd9ac386dea: I wonder if this causes a problem when handling a block template without a witness commitment.

If so, I might just have stratum v2 templates always set a coinbase witness.

ryanofsky replied:
> [8523d19](https://github.com/bitcoin/bitcoin/commit/8523d19bb3f8c982c6e01a4635c98cd9ac386dea): I wonder if this causes a problem when handling a block template without a witness commitment.
> 
> If so, I might just have stratum v2 templates always set a coinbase witness.

This just sets a serialization flag called `allow_witness` and `fAllowWitness` in [`SerializeTransaction`](https://github.com/bitcoin/bitcoin/blob/1a41e63575986887ae34993b4433ec711ae0ffbc/src/primitives/transaction.h#L256) that includes witness data when it is present instead of omitting it. It ensures that if you pass a CTransaction parameter over IPC, the value of the parameter will be unchanged after it is passed. I should add a comment explaining this better, since more serialization flags are added here later in #10102, and may have to be added in the future. But I don't think there should be a particular concern for the mining interface.

sedited reviewed (approved):
ACK c538ec69f266b51c893a374a4bb82796ede3d7cb

ryanofsky commented:
Rebased c538ec69f266b51c893a374a4bb82796ede3d7cb -> 083b886bc6b98fab53e2c6b7dc99a017daa9726e ([`pr/mine-types.4`](https://github.com/ryanofsky/bitcoin/commits/pr/mine-types.4) -> [`pr/mine-types.5`](https://github.com/ryanofsky/bitcoin/commits/pr/mine-types.5), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/mine-types.4-rebase..pr/mine-types.5)) updating for cmake

TheCharlatan commented on src/test/CMakeLists.txt:
```diff
@@ -159,14 +159,6 @@ if(ENABLE_WALLET)
-    ipc_test.cpp
-  )
-
-  target_capnp_sources(bitcoin_ipc_test ${PROJECT_SOURCE_DIR}
```
Did you move these out of here because you'd otherwise get the following error:
```
*** Uncaught exception ***
capnp/compiler/node-translator.c++:1374: context: member.name = arg
kj/filesystem.c++:315: failed: expected parts.size() > 0 [0 > 0]; can't use ".." to break out of starting directory
stack: 59a492c93c94 59a492c93fc7 59a492c916b3 59a492a192f8 59a492b094db 59a492b08eb6 59a492b35fad 59a492b431c8 59a492b043ad 59a492b085a9 59a492b07e69 59a492b080d7 59a492b35b7f 59a492b36834 59a492b4302c 59a492b43495 59a492b587a6 59a492b5556b 59a492b4a2e1 59a492b42aa6 59a492b3fd6c 59a492b3ee13 59a492b1d565 59a492b1710c 59a492b05b7f 59a492b06909 59a492b06c1c 59a492b0bb7e 59a492b0c5b9 59a492a243e9 59a492a5aa22
```

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/30510#discussion_r1747656292

> Did you move these out of here because you'd otherwise get the following error:

Yes, that's the reason. It doesn't seem to be possible to access `src/ipc/` files from the files in the `src/test/` directory by writing `../ipc/` when the `capnp compile` source prefix is `src/test/`, but it does work if the source prefix is `src/`.

I think a different way to work around this would be to pass `--import-path=/path/to/src` to `capnp compile` and then change `using Mining = import "../ipc/capnp/mining.capnp"` to `using Mining = import "/ipc/capnp/mining.capnp", but this would probably be more confusing.

TheCharlatan replied:
That does not really seems less clear to me, but maybe I'm missing something. Using the `--import-path` argument seems akin to the `-I` option I'd pass to a compiler.

ryanofsky replied:
I just think if you see:

```capnp
using Mining = import "../ipc/capnp/mining.capnp";
```

it is very obvious what file is being imported while if you see

```capnp
using Mining = import "/ipc/capnp/mining.capnp"
```

it's less clear what file is being imported because you don't know what the root path is. This is also inconsistent with standard uses of the root path which have package prefixes "capnp" for "mp":

```capnp
using Cxx = import "/capnp/c++.capnp";
using Proxy = import "/mp/proxy.capnp"
```

We would also be polluting the root namespace with our own files (literally every file in src/) so could cause conflict with other files that are being imported.

If we could include a package name like:

```c++
using Mining = import "/bitcoin/ipc/capnp/mining.capnp"
```

This approach would seem more reasonable to me, but I don't think there is a way to do that with our current directory structure. I think just using relative paths is fine here and probably the way to go.

ryanofsky commented:
This is a confusing CI failure:

```
-- Installing: /ci_container_base/ci/scratch/out/share/man/man1/bitcoin-tx.1
-- Installing: /ci_container_base/ci/scratch/out/share/man/man1/bitcoin-cli.1
-- Installing: /ci_container_base/ci/scratch/out/share/man/man1/bitcoind.1
-- Installing: /ci_container_base/ci/scratch/out/share/man/man1/bitcoin-qt.1
-- Installing: /ci_container_base/ci/scratch/out/share/man/man1/bitcoin-util.1
+ false
```

https://cirrus-ci.com/task/6743063731634176


It's also weird because followup PR #30437 which is based on this and includes the same changes shows this test passing.

TheCharlatan commented:
> This is a confusing CI failure:

I'm guessing you saw the actual error further up?
```
In file included from /ci_container_base/src/test/ipc_test.cpp:7:
/ci_container_base/ci/scratch/build-i686-pc-linux-gnu/src/test/ipc_test.capnp.h:18:10: fatal error: '../ipc/capnp/mining.capnp.h' file not found
   18 | #include "../ipc/capnp/mining.capnp.h"
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
[  5%] Building CXX object CMakeFiles/minisketch.dir/src/minisketch/src/fields/generic_4bytes.cpp.o
[  5%] Building CXX object src/zmq/CMakeFiles/bitcoin_zmq.dir/zmqnotificationinterface.cpp.o
[  5%] Building CXX object CMakeFiles/minisketch.dir/src/minisketch/src/fields/generic_5bytes.cpp.o
[  5%] Building CXX object CMakeFiles/minisketch.dir/src/minisketch/src/fields/generic_6bytes.cpp.o
[  5%] Building CXX object CMakeFiles/leveldb.dir/src/leveldb/table/merger.cc.o
[  5%] Building CXX object CMakeFiles/minisketch.dir/src/minisketch/src/fields/generic_7bytes.cpp.o
[  6%] Building CXX object CMakeFiles/minisketch.dir/src/minisketch/src/fields/generic_8bytes.cpp.o
[  6%] Building CXX object CMakeFiles/leveldb.dir/src/leveldb/table/table.cc.o
[  6%] Linking CXX static library libminisketch.a
[  6%] Building CXX object CMakeFiles/leveldb.dir/src/leveldb/table/table_builder.cc.o
[  6%] Built target minisketch
[  6%] Generating capnp/mining.capnp.c++, capnp/mining.capnp.h, capnp/mining.capnp.proxy-client.c++, capnp/mining.capnp.proxy-types.h, capnp/mining.capnp.proxy-server.c++, capnp/mining.capnp.proxy-types.c++, capnp/mining.capnp.proxy.h
[  7%] Generating capnp/echo.capnp.c++, capnp/echo.capnp.h, capnp/echo.capnp.proxy-client.c++, capnp/echo.capnp.proxy-types.h, capnp/echo.capnp.proxy-server.c++, capnp/echo.capnp.proxy-types.c++, capnp/echo.capnp.proxy.h
1 error generated.
make[2]: *** [src/CMakeFiles/bitcoin_ipc_test.dir/build.make:98: src/CMakeFiles/bitcoin_ipc_test.dir/test/ipc_test.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:1036: src/CMakeFiles/bitcoin_ipc_test.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
[  7%] Generating capnp/init.capnp.c++, capnp/init.capnp.h, capnp/init.capnp.proxy-client.c++, capnp/init.capnp.proxy-types.h, capnp/init.capnp.proxy-server.c++, capnp/init.capnp.proxy-types.c++, capnp/init.capnp.proxy.h
[  7%] Building CXX object src/zmq/CMakeFiles/bitcoin_zmq.dir/zmqpublishnotifier.cpp.o
[  7%] Building CXX object CMakeFiles/leveldb.dir/src/leveldb/table/two_level_iterator.cc.o
[  7%] Building CXX object CMakeFiles/leveldb.dir/src/leveldb/util/arena.cc.o
[  7%] Building CXX object src/ipc/CMakeFiles/bitcoin_ipc.dir/capnp/mining.cpp.o
[  7%] Building CXX object CMakeFiles/leveldb.dir/src/leveldb/util/bloom.cc.o
[  7%] Building CXX object CMakeFiles/leveldb.dir/src/leveldb/util/cache.cc.o
```

ryanofsky commented:
re: https://github.com/bitcoin/bitcoin/pull/30510#issuecomment-2334764375

> I'm guessing you saw the actual error further up?

No thanks. I missed that. That is also confusing but at least it's something to work with. I'm sure why it doesn't happen in normal non-CI builds or seem to happen in the followup PR #30437. Will try to reproduce and fix

ryanofsky commented:
Oh, I guess this is failing because it is a parallel build and cmake does not know the generate `src/test/ipc_test.capnp.h` file depends on the generated `ipc/capnp/mining.capnp.h` file, so can fail if these do not happen to be built in the right order. Will need to annotate that somehow.

ryanofsky committed:
build: Make bitcoin_ipc_test depend on bitcoin_ipc

This change is needed to allow generated capnp code in src/ipc/capnp/ to be
used in unit tests for better test coverage in upcoming commits.

ryanofsky reviewed (commented):
Rebased 61c7ad56818888b2ee07db02134456aaf3e9aedc -> 9c98c42a0166e9e201f6e9d32a0692fad5a185f0 ([`pr/mine-types.8`](https://github.com/ryanofsky/bitcoin/commits/pr/mine-types.8) -> [`pr/mine-types.9`](https://github.com/ryanofsky/bitcoin/commits/pr/mine-types.9), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/mine-types.8-rebase..pr/mine-types.9)) adding suggested cmake comment, fixing depends package which pointed at an old libmultiprocess version causing https://cirrus-ci.com/task/4659675541536768, and resolving conflicts with #30509

Sjors commented on src/CMakeLists.txt:
```diff
@@ -325,6 +325,14 @@ if(WITH_MULTIPROCESS)
+  target_capnp_sources(bitcoin_ipc_test ${PROJECT_SOURCE_DIR}
+    test/ipc_test.capnp
+  )
+  add_dependencies(bitcoin_ipc_test bitcoin_ipc_headers)
```
I wasn't using depends, so had to update multiprocess on my system.

sedited reviewed (approved):
ACK 9c98c42a0166e9e201f6e9d32a0692fad5a185f0

Sjors commented:
I successfully tested this code as part https://github.com/Sjors/bitcoin/pull/48#issuecomment-2342745012

Sjors commented on src/CMakeLists.txt:
```diff
@@ -328,6 +328,20 @@ if(WITH_MULTIPROCESS)
+  # reference capnp files in src/ipc/capnp/ by relative path. The Cap'n Proto
+  # compiler only allows importing by relative path when the importing and
+  # imported files are underneath the same compilation source prefix, so the
+  # source prefix must be src/, not src/test/
```
9378b30774d40c8a093c69d2b1807c9eea32455a: I think you need to wrap this in `if(BUILD_TESTS)`

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/30510#discussion_r1757005855

Thanks, added `if(BUILD_TESTS)`

ryanofsky reviewed (commented):
Updated 9c98c42a0166e9e201f6e9d32a0692fad5a185f0 -> 1be749c771cd9fd80361ebb69c87482920b25cd1 ([`pr/mine-types.9`](https://github.com/ryanofsky/bitcoin/commits/pr/mine-types.9) -> [`pr/mine-types.10`](https://github.com/ryanofsky/bitcoin/commits/pr/mine-types.10), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/mine-types.9..pr/mine-types.10)) skipping test library build when tests are disabled, as suggested

itornaza reviewed (approved):
ACK 1be749c771cd9fd80361ebb69c87482920b25cd1

Built and run all tests including the extended functional tests and all of them pass locally. Went through the code changes and read all comments of the pr mostly to educate myself better on the ipc domain. Thanks to all the reviewers and the author for being so explicit.

ryanofsky commented on src/ipc/capnp/common-types.h:
```diff
@@ -103,6 +139,39 @@ decltype(auto) CustomReadField(TypeList<UniValue>, Priority<1>, InvokeContext& i
+//! CustomReadField function that can read from ::capnp::ArrayPtr into
+//! std::vector.
+template <typename LocalType, typename Value, typename Output>
+void CustomBuildField(
```
> The argument list is really difficult to read due to the nature of the arguments. However, I understand that it can be no better than that anyway!

To be fair, I think this could be significantly simpler now that we are using c++20, which supports [concepts](https://en.cppreference.com/w/cpp/language/constraints). In general there should be no reason to use std::enable_if anymore now that concepts are available. There's a lot of older code like this that could be switched to use concepts, and I haven't tried to take that on yet, but this could be a good place to start Thanks for pointing out the issue, and I'll see any there are any simple improvements that can be made here.

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/30510#discussion_r1757372497

Removed enable_if so this should be simpler now. Thanks for bringing this up!

ryanofsky committed:
multiprocess: update common-types.h to use C++20 concepts

Idea came from review comment by ion-
https://github.com/bitcoin/bitcoin/pull/30510#discussion_r1757372497

ryanofsky reviewed (commented):
Updated 1be749c771cd9fd80361ebb69c87482920b25cd1 -> b95bb2179610183d9398d50d8c8fd24b1450ad4d ([`pr/mine-types.10`](https://github.com/ryanofsky/bitcoin/commits/pr/mine-types.10) -> [`pr/mine-types.11`](https://github.com/ryanofsky/bitcoin/commits/pr/mine-types.11), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/mine-types.10..pr/mine-types.11)) switching to concepts instead of enable_if

itornaza reviewed (approved):
ACK b95bb2179610183d9398d50d8c8fd24b1450ad4d

The changes are on `src/ipc/capnp/common-types.h` and greatly enhance code reuse, readability and clarity through the use of concepts. As per the standard review procedure, I rebuilt and run all the tests locally and all of them pass.

hebasto commented on src/CMakeLists.txt:
```diff
@@ -328,6 +328,22 @@ if(WITH_MULTIPROCESS)
+  if(BUILD_TESTS)
+    # bitcoin_ipc_test library target is defined here in src/CMakeLists.txt
+    # instead of src/test/CMakeLists.txt so capnp files in src/test/ are able to
+    # reference capnp files in src/ipc/capnp/ by relative path. The Cap'n Proto
```
Why does referencing `../src/ipc/capnp/` not work from `src/test/CMakeLists.txt`?

Sjors commented:
This probably has a silent merge conflict with #30440.

Sjors commented on src/ipc/capnp/mining.capnp:
```diff
@@ -0,0 +1,37 @@
+    isTestChain @0 (context :Proxy.Context) -> (result: Bool);
+    isInitialBlockDownload @1 (context :Proxy.Context) -> (result: Bool);
+    getTipHash @2 (context :Proxy.Context) -> (result: Data);
+    createNewBlock @3 (context :Proxy.Context, scriptPubKey: Data, options: BlockCreateOptions) -> (result: CBlockTemplate);
```
02c6a1ed58867fa14b1e0c42f476ef45e64ea7ea `-> (result: BlockTemplate);`

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/30510#discussion_r1762668345

> [02c6a1e](https://github.com/bitcoin/bitcoin/commit/02c6a1ed58867fa14b1e0c42f476ef45e64ea7ea) `-> (result: BlockTemplate);`

Thanks, should be updated in latest push.

Sjors commented:
bdfeb12eec31eb885479eed293fe3677cfac52b2 is a really nice simplification.

ryanofsky reviewed (commented):
Rebased b95bb2179610183d9398d50d8c8fd24b1450ad4d -> f9245e4bb880a6d76bbb994a6e1907c6f8c9f205 ([`pr/mine-types.11`](https://github.com/ryanofsky/bitcoin/commits/pr/mine-types.11) -> [`pr/mine-types.12`](https://github.com/ryanofsky/bitcoin/commits/pr/mine-types.12), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/mine-types.11-rebase..pr/mine-types.12)) due to silent conflict with #30440

ryanofsky commented:
Rebased f9245e4bb880a6d76bbb994a6e1907c6f8c9f205 -> 54611e99f5009def3a4559874823ed7fd91c9252 ([`pr/mine-types.12`](https://github.com/ryanofsky/bitcoin/commits/pr/mine-types.12) -> [`pr/mine-types.13`](https://github.com/ryanofsky/bitcoin/commits/pr/mine-types.13), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/mine-types.12-rebase..pr/mine-types.13)) due to conflict with #30875

ryanofsky reviewed (commented):
Updated 54611e99f5009def3a4559874823ed7fd91c9252 -> d59f5a31a1b888051eb47e2f8a5fb8d901de1d10 ([`pr/mine-types.13`](https://github.com/ryanofsky/bitcoin/commits/pr/mine-types.13) -> [`pr/mine-types.14`](https://github.com/ryanofsky/bitcoin/commits/pr/mine-types.14), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/mine-types.13..pr/mine-types.14)) addressing review comments to clarify code

Sjors commented on src/ipc/capnp/common-types.h:
```diff
@@ -16,32 +16,16 @@
-public:
-    static constexpr bool value = decltype(test<T>(nullptr))::value;
+concept Serializable = requires(T t) {
+    decltype(t.Serialize(std::declval<std::nullptr_t&>()))();
```
974e041b794072395e41701cd7a71ebdcc362e48: `serialize.h` already defines a `concept Unserializable`, introduced by @maflcko in #29056.

Though the concept is defined is defined differently there, and either version creates a compiler error when used in both places.

Sjors replied:
Agree it can wait for a followup, but if you have to retouch it would be good to add a comment.

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/30510#discussion_r1766639028

Updated now to use concepts from serialiize.h instead of defining new ones

Sjors commented:
utACK d59f5a31a1b888051eb47e2f8a5fb8d901de1d10

I previously tested this with https://github.com/Sjors/bitcoin/pull/48, will try to do that again.

Would be nice to get a unified `concept Serializable`, see above.

For the most part it should be quite straightforward for other developers in the future to change `src/ipc/capnp/mining.capnp` when something changes in the Mining interface. I've been doing that myself.

The only part that seems (slightly) more involved is in 00a82f45004a887c2b58ed8da3db9280e0ec8b5c. If someone were to refactor `BlockValidationState`, they would have to also adjust its `CustomBuildMessage` and `CustomReadMessage` counterparts.

ryanofsky reviewed (commented):
re: https://github.com/bitcoin/bitcoin/pull/30510#issuecomment-2361045807

Thanks for the review!

> The only part that seems (slightly) more involved is in [00a82f4](https://github.com/bitcoin/bitcoin/commit/00a82f45004a887c2b58ed8da3db9280e0ec8b5c). If someone were to refactor `BlockValidationState`, they would have to also adjust its `CustomBuildMessage` and `CustomReadMessage` counterparts.

I think it might be good idea to change `testBlockValidity` method anyway to return validation result in simpler form than the [`BlockValidationState`](https://github.com/ryanofsky/bitcoin/blob/d59f5a31a1b888051eb47e2f8a5fb8d901de1d10/src/ipc/capnp/mining.capnp#L40-L45) struct. For example, we've been discussing calling the Mining interface from rust code that doesn't have access to the corresponding c++ class and enums. Probably there is something that can be returned that would be more straightforward for rust code to interpret than `mode` and `result` ints and `rejectReason` / `debugMessage` strings.

Sjors commented:
I rebased https://github.com/Sjors/bitcoin/pull/48 on top of the changes here. Managed to mine a testnet4 block.

itornaza reviewed (approved):
re ACK d59f5a31a1b888051eb47e2f8a5fb8d901de1d10

ryanofsky commented on src/ipc/capnp/mining.cpp:
```diff
@@ -0,0 +1,47 @@
+    } else if (reader.getMode() == 1) {
+        dest.Invalid(static_cast<BlockValidationResult>(reader.getResult()), reader.getRejectReason(), reader.getDebugMessage());
+    } else if (reader.getMode() == 2) {
+        assert(reader.getResult() == 0);
+        dest.Error(reader.getRejectReason());
+        assert(reader.getDebugMessage().size() == 0);
```
re: https://github.com/bitcoin/bitcoin/pull/30510#discussion_r1771936209
 
> There are some places where a `BlockValidationState` is first set to invalid with a corresponding result and debug message, before then being set to error later. Presumably if such a `BlockValidateState` were serialized then deserialized, these assertions would be hit. I don't know whether that is actually an issue right now, but it does seem like a potential trap that we might stumble into later.

I guess a way to handle this problem would be to change the corresponding CustomBuildMessage to not set the debug message when state isn't supposed to contain a debug message. I think ideally though we would change the mining interface to avoid using the `BlockValidationState` type and instead use a simpler state representation as suggested https://github.com/bitcoin/bitcoin/pull/30510#pullrequestreview-2315917440

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/30510#discussion_r1771936209

Added a comment to mining.capnp about getting rid of BlockValidationState here and returning something simpler from testBlockValidity

ryanofsky replied:
> re: [#30510 (comment)](https://github.com/bitcoin/bitcoin/pull/30510#discussion_r1771936209)

cca5993b26e6223af31fe1ef5bf8a319cb87cf93 from #31981 finally removes this fragile code after testBlockValidity was dropped in #31196

ryanofsky reviewed (commented):
I think I need to rebase this now that #30409 is merged, so I'll also work on improving comments here as suggested in recent reviews.

Sjors commented:
> I think I need to rebase this now that https://github.com/bitcoin/bitcoin/pull/30409 is merged

Yes

ryanofsky commented on src/ipc/capnp/common-types.h:
```diff
@@ -122,6 +122,36 @@ decltype(auto) CustomReadField(TypeList<UniValue>, Priority<1>, InvokeContext& i
+//! CustomReadField functions for types like PKHash.
+//!
+//! For the LocalType = vector<char> case, it's also not necessary to
+//! have ::capnp::Data CustomReadField function corresponding to this
```
re: https://github.com/bitcoin/bitcoin/pull/30510#discussion_r1771916488

I rewrote the comment and tweaked implementation of this to avoid use of non-standard Span class, so this could be dropped and moved to libmultiprocess in the future. But leaving that for a followup since it will require other changes to libmultiprocess.

ryanofsky commented on src/ipc/capnp/common-types.h:
```diff
@@ -122,6 +122,36 @@ decltype(auto) CustomReadField(TypeList<UniValue>, Priority<1>, InvokeContext& i
+//! that works using Spans, because the bitcoin classes in the codebase like
+//! BaseHash and blob_blob that can converted /to/ Span don't currently have
+//! Span constructors that allow them to be constructed /from/ Span. If they
+//! did, it would simplify things. For example, a generic CustomReadField
```
re: https://github.com/bitcoin/bitcoin/pull/30510#discussion_r1771915747

Rewrote the comment so hopefully it is less confusing now.

ryanofsky reviewed (commented):
Rebased d59f5a31a1b888051eb47e2f8a5fb8d901de1d10 -> 3499c2ca4ba5ae70cebe84614498cc105e208f3d ([`pr/mine-types.14`](https://github.com/ryanofsky/bitcoin/commits/pr/mine-types.14) -> [`pr/mine-types.15`](https://github.com/ryanofsky/bitcoin/commits/pr/mine-types.15), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/mine-types.14-rebase..pr/mine-types.15)) to fix silent conflict with #30409. Also made some improvements in response to recent comments.
Updated 3499c2ca4ba5ae70cebe84614498cc105e208f3d -> fbed31494af9acf6bd543801143a12399b2c6a1a ([`pr/mine-types.15`](https://github.com/ryanofsky/bitcoin/commits/pr/mine-types.15) -> [`pr/mine-types.16`](https://github.com/ryanofsky/bitcoin/commits/pr/mine-types.16), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/mine-types.15..pr/mine-types.16)) fixing up some new comments.
Updated fbed31494af9acf6bd543801143a12399b2c6a1a -> 1a332817665f77f55090fa166930fec0e5500727 ([`pr/mine-types.16`](https://github.com/ryanofsky/bitcoin/commits/pr/mine-types.16) -> [`pr/mine-types.17`](https://github.com/ryanofsky/bitcoin/commits/pr/mine-types.17), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/mine-types.16..pr/mine-types.17)) making more improvements to comments.

Sjors commented:
re-utACK fbed31494af9acf6bd543801143a12399b2c6a1a

I didn't check if the new `CustomBuildField` and `CustomReadField` to serialize `std::chrono` is identical to that in #30437. If not, then I'll test it next time I rebase https://github.com/Sjors/bitcoin/pull/48.

ryanofsky commented:
Thanks for the review!

> I didn't check if the new `CustomBuildField` and `CustomReadField` to serialize `std::chrono` is identical to that in #30437. If not, then I'll test it next time I rebase [Sjors#48](https://github.com/Sjors/bitcoin/pull/48).

They should be the same except the builder now has static asserts to make sure the capnp number type has enough range to represent the std::chrono duration type (that capnp Float64 type can represent MillisecondsDouble in this case).

sedited reviewed (approved):
ACK 1a332817665f77f55090fa166930fec0e5500727

Nice changes to the serialization code since my last review, re-using the existing concepts is great. I wonder if the Deserializable concept could just live in the `serialize.h` header. It seems like it could be generally useful?

My `clang-format` also wants to indent the `requires` expressions, but I don't think that is necessarily more readable.

itornaza commented:
> My `clang-tidy` also wants to indent the `requires` expressions, but I don't think that is necessarily more readable.

Checking with the C++ Core Guidelines and Stroustrup's books, while they do not mention anything about indenting require statements, they do indent it all through the guide like in this [minimal example](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rt-raise)

itornaza reviewed (approved):
ACK 1a332817665f77f55090fa166930fec0e5500727

Very clever rework to maximize code reuse while using already existing concepts. The code is now far easier to read. Built locally and all unit, and functional tests pass including the extended.

ryanofsky commented:
@Sjors did you re-ack the right commit hash in https://github.com/bitcoin/bitcoin/pull/30510#issuecomment-2373443353? (ACK is a for a commit that was replaced about 12 hours before the comment)

ryanofsky commented:
> I wonder if the Deserializable concept could just live in the `serialize.h` header. It seems like it could be generally useful?

Definitely could be. Would be happy to move it if I need to update this again. Reason I didn't move is after the last update is that it's not really consistent with other concepts there, because it not used in that file, and it doesn't take a stream parameter. But I could move it and make it more consistent.

> 
> My `clang-format` also wants to indent the `requires` expressions, but I don't think that is necessarily more readable.

achow101 commented:
ACK 1a332817665f77f55090fa166930fec0e5500727

Sjors commented:
@ryanofsky looks like I copy-pasted the wrong hash. On my machine I have 1a332817665f77f55090fa166930fec0e5500727 (before fetching today).

Sjors commented:
I'm not sure, I'll do another post-merge look.
